### PR TITLE
feat: add delete camp endpoint

### PR DESCRIPTION
**Pull request by udyding**
## Notion ticket link

[Ticket Name](https://www.notion.so/uwblueprintexecs/Task-Board-S22-1a3e651ab8544192bf0fdcbba04f6d03?p=36046d9832f345aa824e498808614fcc)

## Implementation description
* added delete camp endpoint which takes in a camp id and deletes the camp's form questions, file, and camp sessions
* modified camp session deletion to also delete the associated campers

## Steps to test
1. Create a camp with an associated file, form questions, and camp sessions each with campers
2. Ensure deletion will delete all these fields from the database
3. If a camp session deletion fails, the program shouldn't throw an error in the middle of camp sessions deletions, but rather at the end

## What should reviewers focus on?
* Ensure error handling is correct for camp session deletion

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/typescript/services/implementations/campService.ts [hunk 1]
```diff
@@ -129,6 +129,80 @@ class CampService implements ICampService {
...
+          camp.formQuestions.map(async (formQuestionId) => {
+            await MgFormQuestion.findByIdAndDelete(formQuestionId);
+          }),
+        );
+      }
+    } catch (error: unknown) {
+      // rollback form questions deletion
+      try {
+        camp.formQuestions = oldFormQuestions;
+        // camp.campSessions = oldCampSessions;
```
Comment (jjjding2) [inline backend/typescript/services/implementations/campService.ts:166]:
```suggestion
```

**Review by jjjding2 — APPROVED** [review]
lgtm! could we just add a test for the delete camp endpoint in campService.test.ts?

File: backend/typescript/services/implementations/campService.ts [hunk 2]
```diff
@@ -129,6 +129,80 @@ class CampService implements ICampService {
...
+      if (camp.fileName) {
+        await this.storageService.deleteFile(camp.fileName);
+      }
+    } catch (error: unknown) {
+      Logger.error(`Failed to delete camp. Reason = ${getErrorMessage(error)}`);
+      throw error;
+    }
+
+    // keep copy of form questions
+    const oldFormQuestions = [...camp.formQuestions];
```
[the remark this answers is not in the document]
Comment (annechung) [reply backend/typescript/services/implementations/campService.ts:151]: I think this is for deep copying?

File: backend/typescript/services/implementations/campService.ts [hunk 3]
```diff
@@ -129,6 +129,80 @@ class CampService implements ICampService {
...
+      if (!camp) {
+        throw new Error(`Camp with camp ID ${campId} not found.`);
+      }
+
+      // delete the camp's file, if it exists
+      if (camp.fileName) {
+        await this.storageService.deleteFile(camp.fileName);
+      }
+    } catch (error: unknown) {
+      Logger.error(`Failed to delete camp. Reason = ${getErrorMessage(error)}`);
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:146]:
```suggestion
      Logger.error(`Failed to delete camp photo. Reason = ${getErrorMessage(error)}`);
```

File: backend/typescript/services/implementations/campService.ts [hunk 4]
```diff
@@ -270,21 +345,42 @@ class CampService implements ICampService {
...
       );
       throw error;
     }
+
+    // delete all the campers belonging to the camp session
+    const existingCampers = [...campSession.campers];
+    let numberOfCampersDeleted = 0;
+
+    try {
+      existingCampers.forEach(async (camperId) => {
```
Comment (annechung) [inline backend/typescript/services/implementations/campService.ts:372]: let's use `Promise.all` here too

**Review by jjjding2 — APPROVED** [review]
test looks good!

**Pull request closed without merging**